### PR TITLE
feat: Support Cosign signature

### PR DESCRIPTION
**Pull request by jp-gouin**
Fetch cosign signature for each image /w a new cli option.

Silence error if the signature does not exist

### CODE REVIEW

File: cmd/download.go [hunk 1]
```diff
@@ -86,7 +90,8 @@ func DownloadDataPack(c *cli.Context) error {
 		} else {
 			destImg = fmt.Sprintf("%s/images.tar", dest)
 		}
-		err := containers.DownloadImage(c.StringSlice("image"), destImg)
+		log.Println(destImg)
+		err := containers.DownloadImage(c.StringSlice("image"), dest, c.Bool("signature"))
```
Comment (ccleouf66) [inline cmd/download.go]: Use destImg instead of dest

File: cmd/download.go [hunk 2]
```diff
@@ -122,7 +127,7 @@ func DownloadDataPack(c *cli.Context) error {
 
 			// Downlaod container images
 			destImg := fmt.Sprintf("%s/%s-images.tar", chartDest, chartName)
-			err = containers.DownloadImage(imgList, destImg)
+			err = containers.DownloadImage(imgList, destImg, c.Bool("fetchSignature"))
```
Comment (ccleouf66) [inline cmd/download.go]: use c.Bool("signature")) instead c.Bool("fetchSignature")

File: containers/image.go [hunk 3]
```diff
@@ -5,34 +5,24 @@ import (
...
-	if err != nil {
-		log.Printf("Policy context err.\n")
-		return err
-	}
-	defer policyContext.Destroy()
+func DownloadImage(imgList []string, dest string, fetchSignature bool) error {
 
 	// Create new dest archive
-	aw, err := archive.NewWriter(nil, dest)
+	aw, err := archive.NewWriter(nil, fmt.Sprintf("%s/images.tar", dest))
```
Comment (ccleouf66) [inline containers/image.go]: `dest` replaced by `destImg` in `cmd/download.go` so do not provide the destination filename here juste use `aw, err := archive.NewWriter(nil, dest)`

File: containers/image.go [hunk 4]
```diff
@@ -66,28 +57,97 @@ func DownloadImage(imgList []string, dest string) error {
...
+		log.Printf("Policy context err.\n")
+		return err
+	}
+	defer policyContext.Destroy()
+	// Create systemContext to select os and arch
+	sysCtx := &types.SystemContext{
+		ArchitectureChoice: "amd64",
+		OSChoice:           "linux",
+	}
+	_, err = copy.Image(context.Background(), policyContext, destImgRef, srcImgRef, &copy.Options{
```
Comment (ccleouf66) [inline containers/image.go]: func `copy.Image` return the image manifest : [info](https://github.com/containers/image/blob/v3.0.2/copy/copy.go#L125)
So the result of copy.Image can be returned to be used in order to fetch the image digest for the fetchSignature step.

File: containers/image.go [hunk 5]
```diff
@@ -66,28 +57,97 @@ func DownloadImage(imgList []string, dest string) error {
...
+		ReportWriter: os.Stdout,
+		SourceCtx:    sysCtx,
+	})
+	return err
+}
+
+func fetchDigest(srcImgRef types.ImageReference) (*digest.Digest, error) {
```
Comment (ccleouf66) [inline containers/image.go:132]: this function can be removed, the digest can be fetched from the return of copy.Image().

**Review by ccleouf66 — APPROVED** [review]

File: containers/image.go (hunk 5 above)
Comment (jp-gouin) [reply containers/image.go:132]: Nope the digest from the copy.image is not the same digest has the remote one.

File: containers/image.go (hunk 4 above)
Comment (jp-gouin) [reply containers/image.go]: from my tests the returned digest is different than the digest of the container image in the remote registry

File: cmd/download.go (hunk 1 above)
Comment (jp-gouin) [reply cmd/download.go]: actually dest is required in the DownloadImage to create the directory and store the signature

File: cmd/download.go (hunk 2 above)
Comment (jp-gouin) [reply cmd/download.go]: yep

**Pull request merged**